adding proto wrapper for BeaconState

jmozah opened the pull request:
<
**What type of PR is this?**

> Uncomment one line below and remove others.
>
> Bug fix
 Feature
> Documentation
> Other

**What does this PR do? Why is it needed?**

This PR adds a wrapper object over the different beacon states we have (phase0, altair. merge).
Only proto definitions and generate go files are done.

**Which issues(s) does this PR fix?**


Part of  #9962

**Other notes for review**

terencechain commented on proto/eth/wrapper/beacon_wrapper.proto:
```diff
@@ -0,0 +1,29 @@
+// Copyright 2020 Prysmatic Labs.
```
```suggestion
// Copyright 2021 Prysmatic Labs.
```

terencechain commented on proto/eth/wrapper/beacon_wrapper.proto:
```diff
@@ -0,0 +1,29 @@
+import "proto/eth/v1/beacon_state.proto";
+import "proto/eth/v2/beacon_state.proto";
+
+message WrappedBeaconState {
```
Could you add commentary on what this is intended for?

jmozah replied:
oh sure. Added

nisdas commented on proto/eth/wrapper/beacon_wrapper.proto:
```diff
@@ -0,0 +1,34 @@
+
+
+// WrappedBeaconState is used only inside db/kv package and it there for the sole purpose of
+// storing the state objects efficiently. With Normalizing the fields in the "state" bucket
```
```suggestion
// storing the state objects efficiently. With normalizing the fields in the "state" bucket
```

nisdas commented on proto/eth/wrapper/beacon_wrapper.proto:
```diff
@@ -0,0 +1,34 @@
+// BeaconState object itself.
+message WrappedBeaconState {
+  v2.Version version = 1001;        // Used to identify the state object
+  bytes validatorIndexes = 1002;    // stores the compressed indexes of the validator entry hashes
```
Don't we want to add more data related to the circular buffers(randao mixes, block roots, state roots) ?

jmozah replied:
yes, that would be in another PR. Now its about fixing validator entries.

nisdas reviewed (commented):
I actually realised this is referencing the wrong state type completely. In our repo we use the state from `v1alpha1`

jmozah commented:
> I actually realised this is referencing the wrong state type completely. In our repo we use the state from `v1alpha1`

@nisdas 
good catch. Fixed it as v1alpha1 objects.
Also renamed the dir as storage to make it obvious.

kasey commented on proto/prysm/storage/beacon_storage.proto:
```diff
@@ -0,0 +1,29 @@
+// BeaconState object itself.
+message BeaconStateForStorage {
+  v2.Version version = 1001;        // Used to identify the state object
+  bytes validatorIndexes = 1002;    // stores the compressed indexes of the validator entry hashes
```
I think an outer wrapper with a oneOf could be a good move for future proofing. This would allow us to iterate on the schema of the inner BeaconState and provide value in migrations. For instance, we may change our minds on the design of validatorIndexes, or add additional fields. On that note I would also recommend leaving out validatorIndexes until that implementation is final. So more concretely here is what I'm imagining:
```
message BeaconStateForStorage {
    oneOf state_storage {
        BeaconStateVersionedBytes versioned_bytes = 1;
    }
}
message BeaconStateVersionedBytes {
    v2.Version version = 1;
    bytes state = 2;
}
```
cc @prestonvanloon what do you think?

jmozah replied:
we wanted to keep state as ssz bytes since it was faster to unmarshall.
my original design was to have one of the three state objects.

kasey replied:
Right, in my example `bytes state = 2;` is still ssz bytes. The idea of my suggestion is to put the entire envelope into storage so we don't have a situation where some aspects of the schema are versioned and others aren't.

jmozah replied:
> so we don't have a situation where some aspects of the schema are versioned and others aren't.

For some time, we cannot avoid people having different types of schema (older one, historic state, the new one).
We can offer them to move to "new schema" with a migrate subcommand.
When we go to Prysm 3.0 (merge),  we will not support older types of schema and support only one type.

kasey replied:
That's a good point, will the new storage format only be used after an offline migration has been performed, or does the DB code need to be able to differentiate between the 2 kinds of values somehow? If the latter, do you have a plan for dealing with that?

jmozah replied:
It's the later. You can find out by looking at the migration bucket (which says what migration had been done, used already in code today).

kasey commented on proto/prysm/storage/beacon_storage.proto:
```diff
@@ -0,0 +1,29 @@
+// it also stores some indexes and other metadata fields that cannot be stored in the
+// BeaconState object itself.
+message BeaconStateForStorage {
+  v2.Version version = 1001;        // Used to identify the state object
```
This type makes me a little nervous. It is an enum without an "undefined" zero value. For context, in protobuf it is a best practice to define the 0 index of an enum as undefined so that it can be differentiated from a value that is accidentally empty. `v2` is also a package for API types, importing it into v1alpha1 feels a little out of place. I wonder if we should define a new enum for database storage so that we can follow the best practice for safety.

jmozah replied:
ok for not having 0 as a valid enum.
but I disagree to have multiple versions types in multiple packages. 
like phase0 is 0 in API and 1 in storage package.

kasey replied:
They are distinct strongly typed values in separate packages, and they have different uses. One is used by the API and the other by DB unmarshaling code - different envelopes, different layers of the stack (in other words, you aren't ever going to send the storage format over the wire in the rpc package). So it is probably ok for them to be different.

jmozah replied:
@nisdas what do you think. Should we have a separate version for storage?

nisdas replied:
@jmozah I dont have any issues with that. It maybe better to not mix up API and DB versioning here. The 0 value is a valid point as there would be no way to distinguish it from a zero value otherwise. We only plan to use this in the DB package in any case.

jmozah replied:
> They are distinct strongly typed values in separate packages, and they have different uses.

@kasey I think you have a point. I will use a different version for storage and not start that with 0.

> The 0 value is a valid point as there would be no way to distinguish it from a zero value otherwise.

@nisdas All our other versions do have a valid value for 0.  ex: Phase0. Not sure if this applies there too.

nisdas replied:
> All our other versions do have a valid value for 0. ex: Phase0. Not sure if this applies there too.

One of those is determined via the spec apis, we don't have control over those unfortunately. The other isn't as tied to the specs I think since it is an internal prysm value.

rauljordan commented:
What's the latest status @jmozah ? Are we ready for merge soon?

rauljordan commented:
What's the latest status @jmozah ? Are we ready for merge soon?

jmozah commented:
> What's the latest status @jmozah ? Are we ready for merge soon?

oops, will close it soon.

kasey commented on proto/prysm/storage/BUILD.bazel:
```diff
@@ -0,0 +1,33 @@
+    proto = ":proto",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//proto/eth/v2:go_default_library",
```
I don't see this dep being used in code. Do you need to re-run gazelle?

nisdas replied:
Odd, thought the build would have failed in that case.

jmozah replied:
gazel doesn't properly handle this proto deps. It was handcrafted.
oops. it got bulldozed and merged. I will remove it in some other PR.

kasey reviewed (commented):
Thanks for all the follow-up :) I think this needs a gazelle run to clean up the unused dep, otherwise LGTM!
